### PR TITLE
[WOR-678] Add users to pet-creator policy on billing profile when added to workspace

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [WOR-678](https://broadworkbench.atlassian.net/browse/WOR-678)

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

[WOR-678]: https://broadworkbench.atlassian.net/browse/WOR-678?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala [hunk 1]
```diff
@@ -102,4 +111,47 @@ class MultiCloudWorkspaceAclManager(workspaceManagerDAO: WorkspaceManagerDAO, va
...
+
+    if (newWriterEmails.nonEmpty) {
+      for {
+        workspaceBillingProject <- dataSource.inTransaction(
+          _.rawlsBillingProjectQuery.load(RawlsBillingProjectName(workspaceName.namespace))
+        )
+        workspaceBillingProfileId = workspaceBillingProject
+          .getOrElse(
+            throw new RawlsExceptionWithErrorReport(
+              ErrorReport(StatusCodes.InternalServerError,
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala:132]: It'd be good to get coverage on this error condition.

**Review by aherbst-broad — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala [hunk 2]
```diff
@@ -102,4 +111,47 @@ class MultiCloudWorkspaceAclManager(workspaceManagerDAO: WorkspaceManagerDAO, va
...
   )
+
+  def maybeShareWorkspaceNamespaceCompute(
+    policyAdditions: Set[(SamResourcePolicyName, String)],
+    workspaceName: WorkspaceName,
+    ctx: RawlsRequestContext
+  ): Future[Unit] = {
+    val newWriterEmails = policyAdditions.collect { case (SamWorkspacePolicyNames.writer, email) => email }
+
+    if (newWriterEmails.nonEmpty) {
```
Comment (blakery) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala]: A suggestion for an alternative structure:
```scala
  lazy val workspaceBillingProfileId: Future[UUID] = dataSource.inTransaction(
          _.rawlsBillingProjectQuery.load(RawlsBillingProjectName(workspaceName.namespace)
    )
    .map(_.getOrElse(throw new RawlsExceptionWithErrorReport(/*...*/))
    .billingProfileId
    .map(UUID.fromString(_))
    .getOrElse(throw new RawlsExceptionWithErrorReport(/*...*/))
    
  Future.traverse(newWriterEmails) { email =>
      Future(
          billingProfileManagerDAO.addProfilePolicyMember(
              UUID.fromString(workspaceBillingProfileId),
              ProfilePolicy.PetCreator,
              email,
              ctx
          )
        )
  }
```
This avoids the if with the `else Future.successful()` all the way at the end (which I found a little hard to follow at first glance), since the loop only executes if there are elements in newWriterEmails, and the lazy value is only evaluated if the loop executes at least once.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/RawlsWorkspaceAclManager.scala [hunk 3]
```diff
@@ -119,4 +123,30 @@ class RawlsWorkspaceAclManager(val samDAO: SamDAO)(implicit val executionContext
                            ctx: RawlsRequestContext
   ): Future[Unit] =
     samDAO.removeUserFromPolicy(SamResourceTypeNames.workspace, workspace.workspaceId, policyName, email.value, ctx)
+
+  def maybeShareWorkspaceNamespaceCompute(policyAdditions: Set[(SamResourcePolicyName, String)],
+                                          workspaceName: WorkspaceName,
+                                          ctx: RawlsRequestContext
+  ): Future[Unit] = {
+    val newWriterEmails = policyAdditions.collect { case (SamWorkspacePolicyNames.canCompute, email) =>
```
Comment (blakery) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/RawlsWorkspaceAclManager.scala:131]: This could also be converted to the syntax in `MultiCloudWorkspaceACLManager`, if you feel so inclined.

**Review by blakery — APPROVED** [review]
The alternative syntax I mentioned is entirely at your discretion, if you like it better.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala (hunk 2 above)
Comment (marctalbott) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceAclManager.scala]: I swapped the conditional to `.isEmpty` so the `Future.successful()` will be at the top and make it clearer that it will short circuit
